File: Binary_Tree/100_same_tree/100.cpp
#include "../tree.cpp"
#include <algorithm>
#include <vector>


bool isSame(BinaryTree* p, BinaryTree* q) {
    if (p == nullptr && q == nullptr) {
        return true;
    }
    if ((p == nullptr || q == nullptr) || (p -> val != q -> val)) {
        return false;
    }
    return (isSame(p -> left, q -> left) && isSame(p -> right, q -> right));
}


int main() {
    std::vector<int> array = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    BinaryTree* p = build_tree(array);
    print_tree(p);
    std::cout << " " << std::endl;
    BinaryTree* q = build_tree(array);
    print_tree(q);
    std::cout << " " << std::endl;
    
    std::cout << " is same : " << isSame(p, q) << std::endl;
}

File: Binary_Tree/100_same_tree/100.rs
#[path = "../tree.rs"]
mod tree;


fn is_same(p: &Option<Box<tree::BinaryTree>>, q: &Option<Box<tree::BinaryTree>>) -> bool {

    match (p, q) {
        (None, None) => return true,
        (_, None) | (None, _) => return false,
        (Some(p), Some(q)) => {
            if p.val != q.val {
                return false;
            }
            return is_same(&p.left, &q.left) && is_same(&p.right, &q.right);
        } 
    }
}




fn main() {
    let p = tree::build_tree(&mut vec!(1, 2, 3, 4, 5, 6, 7));
    tree::print_tree(&p, 1);

    let q = tree::build_tree(&mut vec!(1, 2, 3, 4, 5, 6, 7));
    tree::print_tree(&q, 1);

    let result = is_same(&p, &q);
    println!("tree is same : {}", result);

}

